DDF-1363 Generate certificates

ahoffer opened the pull request:
@stustison @lessarderic @AzGoalie 

The envisioned use case is  creating a new certificate that uses the machine's DNS name and is signed by the DDF Demo Certificate Authority. The new cert is installed in the keystore and saved to disk.

The established Java API to accomplish these tasks is flexible, but cumbersome. The idea behind these new classes to make some assumptions that simplify the whole process.  See the test case TestCertificateGeneration.installCertificate for an end-to-end flow of how this works.

FYI. I ended up with an class that simplifies loading objects from PEM files. I later went with an approach that does not rely on PEM files, but an advanced use case (sign a new certificate with some CA other than the demo CA) could make use of the PEM file loader.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/139)
<!-- Reviewable:end -->

lessarderic commented:
A few general comments:
* File paths shouldn't contain dots (.), e.g., ddf.security.certificate.generator should be ddf/security/certificate/generator.
* Make sure you add JavaDoc comments on all classes and public methods. See our code comment guidelines wiki.
* Can we reduce the number of inline comments (//)? We try to keep them to a minimum as they become difficult to maintain and keep up to date, and can generally be replaced with private methods when needed. See our code comment guidelines wiki.

_Code comment guidelines wiki: https://codice.atlassian.net/wiki/display/DDF/Code+Comment+Guidelines._

---

Review status: 0 of 20 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

<sup>**[platform/security/certificate/security-certificate-generator/pom.xml, line 76 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxae2xrjpGT4bFJwZ91)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/pom.xml#L76)):</sup>
Any specific reason we have to do all this here? I would expect guava to be packaged automatically if it's declared as a dependency but I could be wrong.

---

<sup>**[platform/security/certificate/security-certificate-generator/pom.xml, line 152 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaeSeOqLRs3JBSEUue)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/pom.xml#L152)):</sup>
Are those temporary? Since this is brand new code, are we going to get at least 75% code coverage?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java, line 43 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxafLKIzcQJLeWt09iJ)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java#L43)):</sup>
Any reason those are protected and not private?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java, line 77 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaiZlTz4N6a5beuZdy)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java#L77)):</sup>
Do all 3 methods need to be public?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java, line 27 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaihcnjReYKV9aqG2K)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java#L27)):</sup>
Is there value in creating all those exceptions, especially since CertificateException is a checked exception and all those sub-classes will have to be caught and handled by the callers? Couldn't we just create one or two RuntimeExceptions and use those instead? We should also keep those exception classes in separate files and add JavaDoc.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 38 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxatXXXjvfVhBpUTR7X)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L38)):</sup>
Can we move this comment in the JavaDoc and maybe remove the last sentence, or at least the "I"? That way people won't wonder who "I" is :-)

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 43 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaqSH1HJR-p4bYI_q3)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L43)):</sup>
Just a name suggestion. Since this class provides utility methods for both keys and certs, do you think we should name it something like SecurityKeyUtilities or PkiUtilities? That would cover everything from private keys, to public keys to public key certificates.

We should also consider making all the methods non-static so that classes that depend on it can easily mock them out.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 45 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxanIimQwKrQoi47Icm)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L45)):</sup>
BOUNCY_CASTLE instead of BC?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 53 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxanbQsXHzWEZ9F6GPn)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L53)):</sup>
Only 3-4 methods are actually used outside this class. Can we make all the others, like this one, private?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 76 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxargUbIYkFyE9jhcJ5)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L76)):</sup>
Don't think that UnknownHostException can ever be thrown here.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 102 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxavcKDEuPN8pxs4m0Q)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L102)):</sup>
We should catch specific exceptions or RuntimeException instead. Catching Exception "hides" new exceptions that can be added in the future.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 108 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxasW42GxBTwYieQ8VK)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L108)):</sup>
private instead of protected?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 117 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxashLr1NEjhHr-aJpQ)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L117)):</sup>
Can we add a link to the RFC 4514 standard in the JavaDoc comments instead of repeating that information here?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 137 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxavz8ltvvAjPWc2pvq)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L137)):</sup>
Do we need to do any validation on the commonName before adding it?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 188 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaoRzAo8-XyHvr6gs8)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L188)):</sup>
getHostName() is only called from here and feels like an odd fit in the CertificateGeneratorUtilities class. Couldn't we use a private method instead and drop the public method on CertificateGeneratorUtilities?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 3 of 30 files at r2.
Review status: 3 of 20 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

AzGoalie commented:
Review status: 3 of 20 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 42 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbC6BlxBZqqSnQZkYH)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L42)):</sup>
as mentioned above by @lessarderic: BOUNCY_CASTLE instead of BC?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 71 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbFbPNsQ6AlRLihlHq)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L71)):</sup>
If we're forcing users to set the expiration date, then we should mention this in the javadocs.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 152 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbGI4yR1IaVBsjYpMw)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L152)):</sup>
I think moving it to CertificateGenerationUtilities makes sense.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 179 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxb3D1CEQjkgO44z7cA)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L179)):</sup>
repeated 'use case is'

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 124 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxb8NdyZM4xa7pfn_6V)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L124)):</sup>
Can remove this return and let it fall through to the other return.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java, line 49 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxb9_JTwgwxS852xVIS)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java#L49)):</sup>
Should this be protected/private if making this a singleton?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java, line 83 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbALAF7780CHLrHOz3)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java#L83)):</sup>
mistype on throw: 'hrow'

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/SecurityFileFacade.java, line 32 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbByaF0hsIZSZsKyAJ)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/SecurityFileFacade.java#L32)):</sup>
as mentioned above by @lessarderic: BOUNCY_CASTLE instead of BC?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/SecurityFileFacade.java, line 45 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbCOWY6SnlxiefBEut)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/SecurityFileFacade.java#L45)):</sup>
Should we check if filePath is empty also?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

beyelerb commented:
Review status: 3 of 20 files reviewed at latest revision, 29 unresolved discussions, some commit checks failed.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 77 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxk2AoqMrgqvck8wQOX)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L77)):</sup>
Do you want to use getCanonicalHostname to return the FQN if available? Typically what you put in your certificates.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 65 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxk3dMRlclFC7BEF4FW)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L65)):</sup>
I would suggest moving to 2048-bit key length as the default. That seems to be the default for a lot of DDF customers.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 109 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxkUgBArHn1-bBww9AG)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L109)):</sup>
info? warn seems pretty severe - most of the time I would not expect the system property to be set.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 162 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxkVTbv5fm6p5T1MVI0)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L162)):</sup>
Using slf4j you can just do this inline - e.g. LOGGER.warn("Could get retrieve certificate named {}", alias)

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 195 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxkW-YfEmW7bqmXfL4V)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L195)):</sup>
slf4j - use parameterized logging here as well, and below

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

ahoffer commented:
Review status: 3 of 20 files reviewed at latest revision, 28 unresolved discussions, some commit checks pending.

---

<sup>**[platform/security/certificate/security-certificate-generator/pom.xml, line 76 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxae2xrjpGT4bFJwZ91)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/pom.xml#L76)):</sup>
Removed <build> section of pom.xml file. I copied the pom.xml from a similar package and did not realized that section was
unnecessary.

---

<sup>**[platform/security/certificate/security-certificate-generator/pom.xml, line 152 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaeSeOqLRs3JBSEUue)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/pom.xml#L152)):</sup>
Removed.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java, line 43 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxafLKIzcQJLeWt09iJ)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java#L43)):</sup>
No good reason they are protected. I have a long-running personal grudge against private member. Would it be better stle/more idiomatic to make them private? I'll assume there is and just switch them over to private.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java, line 77 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaiZlTz4N6a5beuZdy)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateAuthority.java#L77)):</sup>
No. Currently there is no reason to expose the private key through a public access. Changing the method's visibility to private. 

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java, line 27 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaihcnjReYKV9aqG2K)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java#L27)):</sup>
I felt there was value in those exception because they explained an error with a lot of specificity. The client does not have to rely as heavily on good exception messages. However, I really don't know if that adds value in the context these classes will be used. If the exception hierarchy feels to heavy and over-engineered, it probably is. :-)

Here are several approaches to eliminating the custom exception(s);

-Declare methods to throw the most specific excetions. I avoided this approach because methods like #getPrivateKey could throw bizarre errors, like NoSuchAlgorithm

-Declare methods to only throw generalized exceptions: GeneralSecurityException, IOException, etc. This might be a good way to go. Alternatively, we could use CertificateError instead of GeneralSecurityException.

-Catch all errors, log them , and return null. I do not like this approach because it makes debugging more difficult. I prefer to fail fast and for specific reasons that can be caught. But it is an option.

What approach would you use?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 42 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbC6BlxBZqqSnQZkYH)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L42)):</sup>
Done. Code was inlined.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 65 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxk3dMRlclFC7BEF4FW)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L65)):</sup>
Agreed. Made key length 2048.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 71 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbFbPNsQ6AlRLihlHq)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L71)):</sup>
Good point. I have added Javadoc class comment.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 152 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbGI4yR1IaVBsjYpMw)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L152)):</sup>
Done.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 179 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxb3D1CEQjkgO44z7cA)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L179)):</sup>
Got it.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java, line 188 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaoRzAo8-XyHvr6gs8)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateSigningRequest.java#L188)):</sup>
The cohesion between getting the hostname and a Certificate Signing Request felt weak, like they didn't really belong in the same class. Also, there are potentially many ways to get host name/DNS name and I didn't want to couple it too tightly to the Certificate Signing Request.  On the other hand, I did not want client code to have to call the PkiTools directly, so I left a reference to the it in the Certificate Signing Request.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 109 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxkUgBArHn1-bBww9AG)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L109)):</sup>
Changed to info.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 124 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxb8NdyZM4xa7pfn_6V)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L124)):</sup>
KeyStoreException error is pretty severe and represents things that should have already been validated, like keystore password. Probably better not to catch the error at all.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 162 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxkVTbv5fm6p5T1MVI0)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L162)):</sup>
Sweet

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java, line 195 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxkW-YfEmW7bqmXfL4V)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/KeyStoreFile.java#L195)):</sup>
Done.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java, line 49 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxb9_JTwgwxS852xVIS)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java#L49)):</sup>
Changed to protected.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java, line 83 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbALAF7780CHLrHOz3)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/PemFile.java#L83)):</sup>
Fixed

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/SecurityFileFacade.java, line 45 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxbCOWY6SnlxiefBEut)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/SecurityFileFacade.java#L45)):</sup>
I don't know. Should we?

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 38 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxatXXXjvfVhBpUTR7X)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L38)):</sup>
Made a Javadoc comment.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 43 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaqSH1HJR-p4bYI_q3)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L43)):</sup>
How about PkiTools?

I'm not clear on the static/mocking issue. I should drop by your desk and ask you in person. It's nice not to have to create an instance of the utilities because the utility method don't have to preserve any state.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 45 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxanIimQwKrQoi47Icm)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L45)):</sup>
Inlined it.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 53 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxanbQsXHzWEZ9F6GPn)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L53)):</sup>
I like them as public methods. Though I am not called thing publicly now, I can easily see myself doing so. 

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 76 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxargUbIYkFyE9jhcJ5)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L76)):</sup>
Try deleting the throws clause in the signature and tell me what IntelliJ says.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 77 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxk2AoqMrgqvck8wQOX)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L77)):</sup>
Done.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 102 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxavcKDEuPN8pxs4m0Q)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L102)):</sup>
Changed to GeneralSecurityException

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 108 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxasW42GxBTwYieQ8VK)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L108)):</sup>
Done.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 117 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxashLr1NEjhHr-aJpQ)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L117)):</sup>
Yes. Done.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 137 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxavz8ltvvAjPWc2pvq)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L137)):</sup>
I don't know the answer to that. Do we?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 3 of 20 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java, line 0 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaihcnjReYKV9aqG2K)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java#L0)):</sup>
I prefer the second option. I would keep the CertificateGeneratorException but extend from RuntimeException and use it when the method fails because of something else than some invalid input. For invalid inputs, I would use IllegalArgumentException. I didn't look at all the possible exception scenarios so there might be room for more than one custom exceptions if you think those could be helpful to clients.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 43 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaqSH1HJR-p4bYI_q3)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L43)):</sup>
Name sounds good. Drop by any time :-)

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 53 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxanbQsXHzWEZ9F6GPn)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L53)):</sup>
Fair enough. Since we're keeping it public, do you think we should rename it a little to clearly indicate that the string returned is Base64 encoded and not just a simple string? Same for the other method.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 76 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxargUbIYkFyE9jhcJ5)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L76)):</sup>
:(

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 137 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxavz8ltvvAjPWc2pvq)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L137)):</sup>
I looked at the BouncyCastle API and the JavaDoc doesn't say anything so I'm not sure what they validate. I would assume that null is invalid so we should throw an IllegalArgumentException in that case. @stustison, how much should we assume BouncyCastle validates here?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

stustison commented:
So far this looks good, however, there still needs to be an MBean to expose this functionality to an external interface. The end goal of this thing is to have it hooked up to the keystore editor UI. I don't want to drop all this code in and then have no way to access it. It will never get used. You'll also need to create a secondary ticket to update the keystore editor UI to make use of this other back end functionality. 

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

ahoffer commented:
Review status: 3 of 20 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java, line 27 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaihcnjReYKV9aqG2K)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorException.java#L27)):</sup>
Change the input errors to use the IllegalArgumentException
Change CertificateGenertorException to extend RuntimeException
Remove the static nested classes fro the CertificateGeneratorException

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 0 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxanbQsXHzWEZ9F6GPn)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L0)):</sup>
Rename to "bytesToBase64String" and "base64StringToBytes"

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 0 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-JxaqSH1HJR-p4bYI_q3)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L0)):</sup>
Are you covering mock objects in Tuesday's unit testing tech session? If so, I'll wait until then.

---

<sup>**[platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java, line 0 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/139#-Jxavz8ltvvAjPWc2pvq)** ([raw file](https://github.com/codice/ddf/blob/76cf1135461d3345183908667d6d5ca2977f13e2/platform/security/certificate/security-certificate-generator/src/main/java/ddf.security.certificate.generator/CertificateGeneratorUtilities.java#L0)):</sup>
Validate common name is not null. Log WARN if common name is empty string.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

ahoffer commented:
1. Create MBean to expose functionality (create CA, create CSR, sign CSR)
2. Create ticket to wire up keystore editor and new MBean

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->

ahoffer commented:
What kind of interface is going to work best? 
A service that takes all the required parameters and returns a signed cert and private key? 
Or provide a file path to a keystore as an input and add the signed cert and new key to the file? 
Or just look-up the keystore file and do not ask the client code to provide it?

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/139)
<!-- Sent from Reviewable.io -->
